feat: add docker verdaccio 4 release blog (#92)
verdaccio/website: website/blog/2019-05-13-the-new-docker-image-verdaccio-4.md

---
author: Juan Picado
authorURL: https://twitter.com/jotadeveloper
authorFBID: 1122901551
title: The new docker image for Verdaccio 4
---

> Docker has been for Verdaccio a key part of his success, at the time of this writing we have more than 4 millions and growing rapidly. The image provides an easy installation and integration with other tools, as Kubernetes, or composing them with any other Docker image. 

Reviewer: ```suggestion
Docker has been a key part of success for Verdaccio. At the time of this writing we have more than 4 millions image pulls per month and this number is growing rapidly. The image provides an easy way to use Verdaccio in combination with tools like Kubernetes, Docker Compose or any other container orchestration system, simplifying deployment and integration with an existing infrastructure.
```

Author: I like it, but there are 2 minor issues, this is the final test, could you update?
```
Docker has been a key part of success for Verdaccio. At the time of this writing, we have more than 4 million image pulls and this number is growing rapidly. The image provides an easy way to use Verdaccio in combination with tools like Kubernetes, Docker Compose or any other container orchestration system, simplifying deployment and integration with existing infrastructure.
```

Reviewer: Sure

This article aims to describe what have changed and all the improvements and benefits you will enjoy migrating to latest version.

## What’s new?

### Keep it small

> The new image is three times smaller than the previous, around 150 MB instead 500MB, that’s all. We have optimized using [Multi Stage build](https://medium.com/capital-one-tech/multi-stage-builds-and-dockerfile-b5866d9e2f84), this allow us to exclude dependencies and assets are not required for the runtime.  

Reviewer: ```suggestion
The new image is three times smaller than the previous, shrinking down from 500MB to 150MB. We achieved this level of optimization by using [multi-stage build](https://medium.com/capital-one-tech/multi-stage-builds-and-dockerfile-b5866d9e2f84) which allows to exclude dependencies and assets not required for the runtime.  
```

Author: Here I think would be `excluding` instead `to exclude`.

Reviewer: Thanks, I've updated suggestion

### Environment Variables

> To avoid mistakes we have renamed all environment variables being prefixed with VERDACCIO_. This will avoid future collisions and better understanding of the origin and usage of the variable. Here the full list of the new list of variables available in the new image.

Reviewer: ```suggestion
To avoid mistakes we have renamed all environment variables to be prefixed with `VERDACCIO_`. This will avoid future collisions and give better understanding of the origin of variable. Here is the full list of the new variables available in the new image.
```

Author: Here, two small issues, `and give a better` and `origin of the variable`. Below final text.

To avoid mistakes we have renamed all environment variables to be prefixed with `VERDACCIO_`. This will avoid future collisions and give a better understanding of the origin of the variable. Here is the full list of the new variables available in the new image.

Reviewer: Done

[table]

### Support Arbitrary User IDs

> The previous image runs the container with the verdaccio user and group by default, being the UID created randomly within the image. Some users were experimenting issues since some environment requires usage of arbitrary user IDs for security reasons, for such reason we have introduced the environment variable `VERDACCIO_USER_ID` that allows that. 

Reviewer 2: ```suggestion
The previous image runs the container with the verdaccio user and group by default, being the UID created randomly within the image. Some users were experiencing issues since some environments require the usage of custom user IDs for security reasons. To support this, we have introduced the environment variable `VERDACCIO_USER_ID`.
```

Assumption: It is actually not literally *arbitrary* user IDs but *custom* ones.

Furthermore, other optimizations can be possible, as for instance, define a different username using `VERDACCIO_USER_NAME` and such user won’t have permissions to log in by default.

### Security

> We have followed security recommendations removing write permissions to those locations do not need to be modified, avoiding any kind of modification with the default user provided. 

Reviewer 2: ```suggestion
We have followed security recommendations to remove write permissions to those locations that do not need to be modified for the default user. 
```

Seems to me that saying that you avoid any kind of modification would contradict that you only partly removed write permission.

> For instance, the code is written to `/opt/verdaccio` the verdaccio run user cannot modify the compiled resources nor config, only the /verdaccio/storage volume. The image only assign executable permissions to the binary executable required to run verdaccio.

Reviewer 2: ```suggestion
For instance, the code written to `/opt/verdaccio`. The verdaccio run user cannot modify the compiled resources, nor config. Only the `/verdaccio/storage` volume. The image only assigns executable permissions to the binary executable required to run verdaccio.
```

> If you are not using volumes, the `VERDACCIO_USER_NAME` will have only permissions to write in the storage folder, the source code, the configuration and plugins will be read only.

Reviewer 2: ```suggestion
If you are not using volumes, the `VERDACCIO_USER_NAME` will only have permissions to write in the storage folder and the source code. The configuration and plugins will be read only.
```

> To provide your own configuration file, the recommended way is using Docker volumes as follow.

Reviewer 2: ```suggestion
To provide your own configuration file, the recommended way is using Docker volumes like so:
```

[example volume]

We force a specific uid **10001** for the run user, assign the root group to the locations that need to be written by the run user; if running in a normal environment, the specific id is used and permissions are correct; if running on a randomized uid environment like openshift, the non-existent user gets assigned the root group and allowed write access to relevant locations.

The entrypoint will add the user to `/etc/passwd` in case the user is running as a random uid (openshift). That way, the typical tools like whoami and so can still work. 

## Conclusions

This new image has been tested in production for months and is quite stable, thus there is no fear to give it a try. We have improved in several areas, but still there are a lot to do and for that we need you. If you are DevOps do not hesitate to give us your feedback or contribute directly in discussions and future PRs to take the Verdaccio Docker image to the next level. We count on you.

## Contributions

> We want to thanks to **Diego Louzán**, **Dimitri Koppriwa**, **Sergio Herrera**, [Ben Tucker] (https://github.com/btucker), [Michiel De Mey](https://github.com/MichielDeMey) and me [Juan Picado](https://github.com/juanpicado) for this amazing job improving the Docker image. 

Reviewer: Link to Ben Tucker profile is broken. Also, in my (certainly not expert) opinion, "we want to thank somebody" or "we want to give thanks to somebody" sounds better than "we want to thanks to somebody".

```suggestion
We want to thank **Diego Louzán**, **Dimitri Koppriwa**, **Sergio Herrera**, [Ben Tucker](https://github.com/btucker), [Michiel De Mey](https://github.com/MichielDeMey) and me [Juan Picado](https://github.com/juanpicado) for this amazing job improving the Docker image. 
```
